### ISSUE TITLE
How to retrieve a device's name.

### ISSUE THREAD

**Issue by g-ortuno** [WebBluetoothCG/web-bluetooth#299]
Up until now the implementations have been relying on the platform's provided way of retrieving the name e.g. android: BluetoothDevice.getName(), osx [CBPeripheral name], etc.

We recently found out that BluetoothDevice.getName() is really slow. So I started thinking of how to implement our own name retrieval and wondering if that's something we should spec.

The spec simply [states](https://webbluetoothcg.github.io/web-bluetooth/#get-the-bluetoothdevice-representing):

> 1. If device has a partial or complete Bluetooth Device Name, set result.name to that string.

Should that be more specific? For example what name do we use if the device is advertising one name but generic access has a different name? Should we update the name after connecting to the device? If the user changes the name do we update BluetoothDevice's name?

**Comment by Vudentz** [comment]
I guess it shall be updated once connected but we should be able to use the advertised name,  if one is set, in the meantime. At least with BlueZ we also keep a cache of the devices names.

**Comment by jyasskin** [comment]
Do you know what [`BluetoothDevice.getName()`](https://developer.android.com/reference/android/bluetooth/BluetoothDevice.html#getName()) is doing that's really slow? Its documentation says it's cached, which shouldn't take much time, but if it's actually doing the characteristic read at that point, it could take a long time.

CSSv6 A.1.2.1 says the UA can assume that the advertised name is consistent with the `org.bluetooth.characteristic.gap.device_name` value, and I don't really want to have Web Bluetooth say how to resolve violations of the Bluetooth spec. We could explicitly say that the UA can retrieve `gap.device_name` at any time and use that to set the name field. BT4.2 3.C.12.1 only implies that successfully writing the value updates the value that would be read, but I'd like to behave as if that implication is true. I can mention that in the same place as the permission to read `gap.device_name`.

**Comment by scheib** [comment]
I agree with using a cached advertised name, optionally replacing it with gap.device_name once connected. I don't think we should use any sort of alias (an alternate name that is not provided by a remote device).

**Comment by Emill** [comment]
Well it's impossible to make it perfect as long as the name is a property (unreliable cache) of the device rather than an asynchronous operation (delivering the correct name as the result of a promise).

Let me give you some info about my own experiences:

If we talk about Android, I usually myself never use BluetoothDevice.getName() since I found out it sometimes returns null and I don't really know where it got the name from nor if it is the full name. I rather just get the name by parsing the advertisement data when I'm handling an advertising device, and explicitly read the GATT name characteristic when I'd like to know the name when I'm connected. If there is no name in the advertisement data, or only a partial name, this is the only reliable way to get the full name.

Since it takes some time to read the name, the property and name cache can not be populated directly after the device has been connected. As there is no callback when the name has been populated (and I don't even know when or if Android requests it), I can't know when to fetch it. Therefore I always do my own GATT read. Note that since the name reading "occupies" one round trip and I rather want to do other stuff before reading the name, I don't read the name immediately after I'm connected.

I also don't understand how BluetoothDevice.getName() could be slow. It's just a Binder call to the Bluetooth service asking for a cached name (see https://android.googlesource.com/platform/packages/apps/Bluetooth/+/32a92489e9779d10c613a57d86135a7de6d3d7c4/src/com/android/bluetooth/btservice/AdapterService.java#1132).

Everything also of course depends on how much the developer actually cares about the name, if it's just used for display in some UI or if it is used in some other way. That said, I think it's perfectly fine if the `name` property of the device is the cached advertised name. If the user explicitly reads the name GATT characteristic it could be updated with that value (or if the underlying stack reads it autonomously).
